compiler/rustc_hir_typeck/src/pat.rs

```diff
@@ -746,9 +752,14 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 self.tcx.dcx(),
                 ident.span,
                 E0596,
-                "cannot bind with `ref mut` behind an `&` pattern"
+                "cannot borrow as mutable inside an `&` pattern"
+            );
+            err.span_suggestion(
+                and_pat_span,
+                "replace this `&` with `&mut`",
+                "&mut ",
+                Applicability::MachineApplicable,
```

Comment:
Could you add a comment explaining why this can only happen when the type of the pattern is `&mut T`? At first glance I'd expect it to trigger also in the `let &mut ref x = &0;` case.